protocols/upnp/src/behaviour.rs

```diff
@@ -0,0 +1,515 @@
+// Copyright 2017 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+#![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
+
+use std::{
+    borrow::Borrow,
+    collections::HashMap,
+    error::Error,
+    hash::{Hash, Hasher},
+    net::{Ipv4Addr, SocketAddrV4},
+    pin::Pin,
+    sync::Arc,
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use crate::{
+    provider::{Protocol, Provider},
+    Config,
+};
+use futures::{future::BoxFuture, stream::FuturesUnordered, Future, FutureExt, StreamExt};
+use futures_timer::Delay;
+use libp2p_core::{multiaddr, transport::ListenerId, Endpoint, Multiaddr};
+use libp2p_swarm::{
+    derive_prelude::PeerId, dummy, ConnectionDenied, ConnectionId, ExpiredListenAddr, FromSwarm,
+    NetworkBehaviour, NewListenAddr, PollParameters, ToSwarm,
+};
+
+/// The duration in seconds of a port mapping on the gateway.
+const MAPPING_DURATION: u64 = 3600;
+
+/// Renew the Mapping every half of `MAPPING_DURATION` to avoid the port being unmapped.
+const MAPPING_TIMEOUT: u64 = MAPPING_DURATION / 2;
+
+/// Map a port on the gateway.
+fn map_port<P: Provider + 'static>(
+    gateway: Arc<P::Gateway>,
+    mapping: Mapping,
+    permanent: bool,
+) -> BoxFuture<'static, Event> {
+    let duration = if permanent { 0 } else { MAPPING_DURATION };
```

Comment:
Is there any use case where one might want to have a permanent mapping to the port?

Reply:
This was made prior to https://github.com/libp2p/rust-libp2p/pull/4156#discussion_r1252394720 but after seeing your comment there it made more sense on why it could be an option